### PR TITLE
refactor: localstore transaction

### LINKED ISSUE #4341: replace storer transactions with leveldb batches
the chunkstore operations do not currently accept the leveldb batches for writes.
this can cause issues especially in areas of the localstore where the chunkstore is used in tandem with other stores that does indeed utilizes batches for write (eg reserve.Put).

the chunkstore transaction implementation is too complex and bug prone.
transactions for sharky especially is not needed, if a db operation fails, the written data and used slot can simply remain as is and do not need to be recovered.
we have `compact` cmd now which can give the used sharky space back to the node.

also the `reverseOP` transaction idea for the indexstore should ideally be completely dropped and all writes must be done using the 'leveldb.batch`.

### PR DESCRIPTION
**Pull request by istae**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [x] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues.

### Description
An extensive cleanup of the db transaction.
Every localstore write operation now utilizes a transaction. Calls to the chunkstore are also part of the transaction.

also tackled: evict just enough chunks of a batch to fall below the reserve capacity

TODO:
revert reserve size change
revert reserve wake up duration

### Related Issue (Optional)
closes #4341 #4538 #4636 #4633

### CODE REVIEW

File: pkg/node/node.go [hunk 1]
```diff
@@ -185,8 +185,8 @@ const (
 	minPaymentThreshold           = 2 * refreshRate           // minimal accepted payment threshold of full nodes
 	maxPaymentThreshold           = 24 * refreshRate          // maximal accepted payment threshold of full nodes
 	mainnetNetworkID              = uint64(1)                 //
-	ReserveCapacity               = 4_194_304                 // 2^22 chunks
-	reserveWakeUpDuration         = 30 * time.Minute          // time to wait before waking up reserveWorker
+	ReserveCapacity               = 16384                     // 2^14 chunks, TODO: replace before merging to master
+	reserveWakeUpDuration         = 5 * time.Minute           // time to wait before waking up reserveWorker
```
Comment (acha-bill) [inline pkg/node/node.go:189]: Reminder :)

**Review by acha-bill — APPROVED** [review]

**Review by janos — COMMENTED** [review]
The changeset in this PR is quite large +3481 −5890 lines across 117 files. I am not able to validate all the changes for correctness as I would need more context on them. Some of the changes look to me not related to localstore transactions. I think that an interactive session on this PR would be more efficient than submitting comments. Esad, would you consider doing an interactive session?

**Review by acha-bill — CHANGES_REQUESTED** [review]

File: pkg/storer/internal/cache/cache_test.go [hunk 2]
```diff
@@ -295,63 +249,13 @@ func TestCache(t *testing.T) {
 			}
 		})
 	})
-	t.Run("handle error", func(t *testing.T) {
```
Comment (janos) [inline pkg/storer/internal/cache/cache_test.go:298]: It would be good to keep error checking tests.

File: pkg/storer/internal/chunkstore/chunkstore.go [hunk 3]
```diff
@@ -33,6 +33,175 @@ const RetrievalIndexItemSize = swarm.HashSize + 8 + sharky.LocationSize + 4
...
+type IterateResult struct {
+	Err  error
+	Item *RetrievalIndexItem
+}
+
+// IterateLocations iterates over entire retrieval index and plucks only sharky location.
+func IterateLocations(
+	ctx context.Context,
+	st storage.Reader,
+	locationResultC chan<- LocationResult,
```
Comment (janos) [inline pkg/storer/internal/chunkstore/chunkstore.go:165]: As the goroutine that is created in here takes the full ownership of the locationResultC channel (it writes to it and closes the channel when it it done), it would be better to protect the channel as returning it as the read only channel. In all useasges in this pr, the channel is creaded before the IterateLocations() call and only read from it.

File: pkg/storer/internal/reserve/items.go [hunk 4]
```diff
@@ -104,11 +102,11 @@ func (b *BatchRadiusItem) Unmarshal(buf []byte) error {
...
-	Bin       uint8
-	BinID     uint64
-	Address   swarm.Address
-	BatchID   []byte
-	ChunkType swarm.ChunkType
+	Bin     uint8
+	BinID   uint64
+	Address swarm.Address
+	BatchID []byte
+	Type    swarm.ChunkType
```
Comment (janos) [inline pkg/storer/internal/reserve/items.go:109]: Is this rename really necessary for the new transaction? It adds additional changes that may not be needed.

File: pkg/storer/internal/upload/uploadstore.go [hunk 5]
```diff
@@ -474,168 +465,117 @@ func (u *uploadPutter) Close(s internal.Storage, writer storage.Writer, addr swa
...
-		}
-
-		err = chunkstamp.DeleteWithStamp(writer, chunkStampNamespace, address, stamp)
-		if err != nil {
-			return fmt.Errorf("failed deleting chunk stamp %x: %w", batchID, err)
-		}
-	}
-	return nil
+	var eg errgroup.Group
+	eg.SetLimit(8)
```
Comment (janos) [inline pkg/storer/internal/upload/uploadstore.go:508]: 8 seems like a value that is optimal for some systems, others may be better with different constant. Is there a sense to put the number of CPUs?

File: pkg/storer/internal/transaction/mem.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+package transaction
+
+import (
+	"github.com/ethersphere/bee/v2/pkg/storage"
+	"github.com/ethersphere/bee/v2/pkg/storage/storageutil"
+)
+
+type memcache struct {
+	storage.IndexStore
+	data map[string][]byte
```
Comment (janos) [inline pkg/storer/internal/transaction/mem.go:14]: Be careful with using a map to add a large number of keys as it will internally grow the number of map buckets and even when the keys are deleted the number of buckets will not shrink. This can lead to a "data leak" manifested as more memory used by the map even if it is cleared.

**Review by janos — CHANGES_REQUESTED** [review]
Nice approach on the transaction handling. It is an improvement, for sure. 

The PR is rather large and it needs a lot of attention and with so many changes it is something harder to see if a change is related to the transaction or general code improvement. It would be good to separate code improvements from the new transaction into another PR, but that would require a lot of additional work. It is crucial to have do extensive correctness and load testing on integration.

File: cmd/bee/cmd/db.go [hunk 7]
```diff
@@ -227,6 +230,75 @@ func dbValidatePinsCmd(cmd *cobra.Command) {
...
+			logger.Warning("This is a destructive process. If the process is stopped for any reason, the reserve may become corrupted.")
+			logger.Warning("You have another 10 seconds to change your mind and kill this process with CTRL-C...")
+			time.Sleep(10 * time.Second)
+			logger.Warning("proceeding with repair...")
+
+			d, err := cmd.Flags().GetDuration(optionNameSleepAfter)
+			if err != nil {
+				logger.Error(err, "getting sleep value failed")
+			}
+			defer func() { time.Sleep(d) }()
```
[the remark this answers is not in the document]
Comment (istae) [reply cmd/bee/cmd/db.go:268]: this is a infra requirement, to prevent the pod from restarting when it quits

File: pkg/api/stewardship.go [hunk 8]
```diff
@@ -41,18 +41,7 @@ func (s *Service) stewardshipPutHandler(w http.ResponseWriter, r *http.Request)
...
-		if err != nil {
-			logger.Debug("unable to find old batch for reference", "error", err)
-			logger.Error(nil, "unable to find old batch for reference")
-			jsonhttp.NotFound(w, "unable to find old batch for reference, provide new batch id")
-			return
-		}
-	} else {
-		batchID = headers.BatchID
-	}
+	batchID = headers.BatchID
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/api/stewardship.go:44]: it's no longer possible because the uploadItem no longer exists, the new requirement is that the user simply provides the batchID

File: pkg/storer/storer.go [hunk 9]
```diff
@@ -78,8 +78,6 @@ type UploadStore interface {
 	DeleteSession(tagID uint64) error
 	// ListSessions will list all the Sessions currently being tracked.
 	ListSessions(offset, limit int) ([]SessionInfo, error)
-	// BatchHint will return the batch ID hint for the chunk reference if known.
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/storer/storer.go:81]: done

File: pkg/node/statestore.go [hunk 10]
```diff
@@ -32,10 +32,7 @@ func InitStateStore(logger log.Logger, dataDir string, cacheCapacity uint64) (st
 		return nil, nil, err
 	}
 
-	caching, err := cache.Wrap(ldb, int(cacheCapacity))
-	if err != nil {
-		return nil, nil, err
-	}
+	caching := cache.MustWrap(ldb, int(cacheCapacity))
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/node/statestore.go:35]: it's old artifact, can be removed.

File: pkg/file/joiner/joiner_test.go [hunk 11]
```diff
@@ -1367,3 +1366,70 @@ func TestJoinerRedundancyMultilevel(t *testing.T) {
 		})
 	}
 }
+
+type chunkStore struct {
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/file/joiner/joiner_test.go:1370]: the joiner tests do not expect the mock to not take into account the amount of Put calls before a delete, which is different from the more "correct implementation" of the  inmemchunkstore .

File: pkg/storer/internal/transaction/transaction.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+	store   storage.Reader
+	batch   storage.Batch
+	metrics metrics
+}
+
+func (s *indexTrx) Get(i storage.Item) error           { return s.store.Get(i) }
+func (s *indexTrx) Has(k storage.Key) (bool, error)    { return s.store.Has(k) }
+func (s *indexTrx) GetSize(k storage.Key) (int, error) { return s.store.GetSize(k) }
+func (s *indexTrx) Iterate(q storage.Query, f storage.IterateFn) (err error) {
+	defer handleMetric("iterate", s.metrics)(err)
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/storer/internal/transaction/transaction.go:269]: good catch

File: pkg/node/node.go [hunk 13]
```diff
@@ -176,19 +176,21 @@ type Options struct {
...
-	basePrice                     = 10000                     // minimal price for retrieval and pushsync requests of maximum proximity
+	basePrice                     = 10_000                    // minimal price for retrieval and pushsync requests of maximum proximity
 	postageSyncingStallingTimeout = 10 * time.Minute          //
 	postageSyncingBackoffTimeout  = 5 * time.Second           //
 	minPaymentThreshold           = 2 * refreshRate           // minimal accepted payment threshold of full nodes
 	maxPaymentThreshold           = 24 * refreshRate          // maximal accepted payment threshold of full nodes
 	mainnetNetworkID              = uint64(1)                 //
-	ReserveCapacity               = 4_194_304                 // 2^22 chunks
-	reserveWakeUpDuration         = 30 * time.Minute          // time to wait before waking up reserveWorker
+	ReserveCapacity               = 4194304                   // 2^22 chunks
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/node/node.go:188]: done

File: pkg/puller/puller.go [hunk 14]
```diff
@@ -35,15 +36,19 @@ var errCursorsLength = errors.New("cursors length mismatch")
 const (
 	DefaultHistRateWindow = time.Minute * 15
 
-	intervalPrefix = "sync_interval"
+	IntervalPrefix = "sync_interval"
 	recalcPeersDur = time.Minute * 5
+
+	maxChunksPerSecond = 1000
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/puller/puller.go:42]: its roughly 4 MB/s, it's an arbitrary number, i added comments

File: pkg/storer/internal/cache/cache.go [hunk 15]
```diff
@@ -188,23 +178,87 @@ func (c *Cache) Getter(store internal.Storage) storage.Getter {
...
 func (c *Cache) ShallowCopy(
 	ctx context.Context,
-	store internal.Storage,
+	store transaction.Storage,
 	addrs ...swarm.Address,
 ) (err error) {
 
-	c.glock.Lock()
-	defer c.glock.Unlock()
+	// TODO: add proper mutex locking before usage
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/storer/internal/cache/cache.go:250]: this func is not used anywhere so safe to leave this as is

File: pkg/node/statestore.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (istae) [reply pkg/node/statestore.go:35]: done

File: pkg/puller/puller_test.go [hunk 16]
```diff
@@ -327,42 +328,102 @@ func TestBinReset(t *testing.T) {
...
+		rs:       rs,
+	})
 
-	waitSyncCalledBins(t, pullsync, addr, 1)
+	waitSyncCalledBins(t, pullsync, peerAddr, 0, 1, 2, 3)
+
+	// syncs bin 2 only as this peer is out of depth
+	rs.SetStorageRadius(3)
+	kad.Trigger()
+	time.Sleep(time.Millisecond * 250)
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/puller/puller_test.go:418]: i agree, the puller tests would need a redesign and this is beyond the scope of this PR

File: pkg/retrieval/retrieval.go [hunk 17]
```diff
@@ -371,13 +371,10 @@ func (s *Service) retrieveChunk(ctx context.Context, quit chan struct{}, chunkAd
 
 func (s *Service) prepareCredit(ctx context.Context, peer, chunk swarm.Address, origin bool) (accounting.Action, error) {
 
-	creditCtx, cancel := context.WithTimeout(ctx, time.Second)
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/retrieval/retrieval.go:374]: it should be safe to wait as long as the context does not timeout, the one second is very arbitrary

File: pkg/pushsync/pushsync.go [hunk 18]
```diff
@@ -550,19 +550,15 @@ func (ps *PushSync) pushChunkToPeer(ctx context.Context, peer swarm.Address, ch
 }
 
 func (ps *PushSync) prepareCredit(ctx context.Context, peer swarm.Address, ch swarm.Chunk, origin bool) (accounting.Action, error) {
-
-	creditCtx, cancel := context.WithTimeout(ctx, time.Second)
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/pushsync/pushsync.go:554]: see above comment

File: pkg/storer/internal/cache/cache.go [hunk 19]
```diff
@@ -188,23 +178,87 @@ func (c *Cache) Getter(store internal.Storage) storage.Getter {
...
 	entries := make([]*cacheEntry, 0, len(addrs))
 
 	defer func() {
 		if err != nil {
-			for _, entry := range entries {
-				err = errors.Join(store.ChunkStore().Delete(context.Background(), entry.Address))
-			}
+			_ = store.Run(context.Background(), func(s transaction.Store) error {
+				for _, entry := range entries {
+					err = errors.Join(s.ChunkStore().Delete(context.Background(), entry.Address))
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/storer/internal/cache/cache.go:258]: done

File: pkg/storer/migration/reserveRepair.go [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+		if err != nil {
+			return err
+		}
+
+		logger.Info("counted all batch radius entries", "total_entries", len(batchRadiusItems))
+
+		var missingChunks atomic.Int64
+		var invalidSharkyChunks atomic.Int64
+
+		var bins [32]uint64
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/storer/migration/reserveRepair.go:153]: done

File: pkg/storer/migration/reserveRepair.go [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+			return err
+		}
+
+		err = checkBinIDs()
+		if err != nil {
+			return err
+		}
+
+		batchRadiusCnt, _ := st.IndexStore().Count(&reserve.BatchRadiusItem{})
+		chunkBinCnt, _ := st.IndexStore().Count(&reserve.ChunkBinItem{})
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/storer/migration/reserveRepair.go:229]: done

File: pkg/storer/internal/upload/uploadstore.go (hunk 5 above)
Comment (istae) [reply pkg/storer/internal/upload/uploadstore.go:508]: the idea here is to prevent the loop from creating too many goroutines and overload the cpu/disk

File: cmd/bee/cmd/db.go [hunk 22]
```diff
@@ -227,6 +230,75 @@ func dbValidatePinsCmd(cmd *cobra.Command) {
...
+			if err != nil {
+				return fmt.Errorf("get data-dir: %w", err)
+			}
+			if dataDir == "" {
+				return errors.New("no data-dir provided")
+			}
+
+			logger.Warning("Repair will recreate the reserve entries based on the chunk availability in the chunkstore. The epoch time and bin IDs will be reset.")
+			logger.Warning("The pullsync peer sync intervals are reset so on the next run, the node will perform historical syncing.")
+			logger.Warning("This is a destructive process. If the process is stopped for any reason, the reserve may become corrupted.")
```
[the remark this answers is not in the document]
Comment (istae) [reply cmd/bee/cmd/db.go:259]: added extra logs to warn users to backup the data before running the cmd

File: pkg/api/stewardship.go [hunk 23]
```diff
@@ -29,7 +29,7 @@ func (s *Service) stewardshipPutHandler(w http.ResponseWriter, r *http.Request)
 	}
 
 	headers := struct {
-		BatchID []byte `map:"Swarm-Postage-Batch-Id"`
+		BatchID []byte `map:"Swarm-Postage-Batch-Id" validate:"required"`
```
[the remark this answers is not in the document]
Comment (istae) [reply pkg/api/stewardship.go:32]: without batchHint, the batch ID would be needed unfortunately.

File: pkg/storer/internal/cache/cache_test.go (hunk 2 above)
Comment (istae) [reply pkg/storer/internal/cache/cache_test.go:298]: done

File: pkg/storer/internal/reserve/items.go (hunk 4 above)
Comment (istae) [reply pkg/storer/internal/reserve/items.go:109]: sure, reverted

File: pkg/storer/internal/chunkstore/chunkstore.go (hunk 3 above)
Comment (istae) [reply pkg/storer/internal/chunkstore/chunkstore.go:165]: done

File: pkg/storer/internal/transaction/mem.go (hunk 6 above)
Comment (istae) [reply pkg/storer/internal/transaction/mem.go:14]: the file is not needed at this point, removed

File: pkg/puller/puller_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (janos) [reply pkg/puller/puller_test.go:418]: Cool, it would be good to have an issue regarding correcting tests approaches like this.

File: pkg/storer/internal/upload/uploadstore.go (hunk 5 above)
Comment (janos) [reply pkg/storer/internal/upload/uploadstore.go:508]: Exactly because of this. Value 8 may be good for some system, while other values can be good for other systems. Like where is only one CPU, the value of 1 would probably be better than 8. That is why I've suggested to set the number of CPUs instead the fixed value of 8.

File: cmd/bee/cmd/db.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (acha-bill) [reply cmd/bee/cmd/db.go:269]: We should add a comment then.

File: pkg/storer/internal/upload/uploadstore.go (hunk 5 above)
Comment (istae) [reply pkg/storer/internal/upload/uploadstore.go:508]: aha I see what you mean, then i will use `runtime.NumCPU()`

**Review by janos — APPROVED** [review]
Thanks for addressing the comments.
